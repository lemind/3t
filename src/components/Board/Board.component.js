import React from 'react';
import './Board.less';

import { SquareComponent as Square } from './../Square/Square.component';
import AI from './../../helpers/AI'

export class BoardComponent extends React.Component {
  constructor(props) {
    super();

    const squares = []
    for (var i = 0; i < 9; i++) {
      squares.push(i)
    }

    this.state = {
      squares: props.process.squares,
      playerSign: props.process.playerSign,
      AISign: props.process.AISign,
      finished: props.process.finished,
      winner: props.process.winner
    };

    this.actions = {};
    this.actions.resetGame = props.resetGame;
    this.actions.updateBoard = props.updateBoard;
    this.actions.setWinner = props.setWinner;
  }

  isEmptySquare (key) {
    return Number.isInteger(this.state.squares[key])
  }

  handleClick(i) {
    if (this.state.finished || !this.isEmptySquare(i)) return
    const squares = this.state.squares.slice();
    squares[i] = this.state.playerSign;
    this.setState({ squares },
      this.checkWinner.bind(this, this.state.playerSign)
    );
  }

  checkWinner(player) {
    if (this.isWinner(player)) {
      this.finishGame(player)
    } else {
      if (player === this.state.playerSign) {
        this.AIMove()
      }
    }
  }

  AIMove() {
    const squares = this.state.squares.slice();
    const AIsign = this.state.AISign
    const bestIndex = AI.getBestMove(squares, AIsign).index
    squares[bestIndex] = AIsign;
    this.setState({ squares },
      this.checkWinner.bind(this, this.state.AISign)
    );
  }

  isWinner(player) {
    return AI.isWin(this.state.squares, player)
  }

  finishGame(winner) {
    this.setState({
      finished: true,
      winner
    });
  }

  renderSquare(i) {
    return (
      <Square
        value={this.state.squares[i]}
        onClick={() => this.handleClick(i)}
      />
    );
  }

  static getDerivedStateFromProps(nextProps, prevState) {
    return {
      squares: nextProps.process.squares,
      finished: nextProps.process.finished,
      winner: nextProps.process.winner
    }
  }

  render() {
    return (
      <div className="board">
        <div className="board__row">
          {this.renderSquare(0)}
          {this.renderSquare(1)}
          {this.renderSquare(2)}
        </div>
        <div className="board__row">
          {this.renderSquare(3)}
          {this.renderSquare(4)}
          {this.renderSquare(5)}
        </div>
        <div className="board__row">
          {this.renderSquare(6)}
          {this.renderSquare(7)}
          {this.renderSquare(8)}
        </div>
      </div>
    )
  };
}
