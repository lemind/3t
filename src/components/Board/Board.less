.board {
  margin-top: 20px;
  &__row:after {
    clear: both;
    content: "";
    display: table;
  }
}
