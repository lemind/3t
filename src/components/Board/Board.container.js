import { connect } from 'react-redux';

import { BoardComponent } from './Board.component';
import { processActions } from '../../redux/process/index';

export const BoardContainer = connect(
  function mapStateToProps(state) {
    return {
      process: state.process
    };
  },
  function mapDispatchToProps(dispatch) {
    return {
      updateBoard: (board) => {
        dispatch(processActions.updateBoard(board))
      },
      setWinner: (winner) => {
        dispatch(processActions.setWinner(winner))
      },
      startGame: () => {
        dispatch(processActions.startGame())
      }
    };
  }
)(BoardComponent);