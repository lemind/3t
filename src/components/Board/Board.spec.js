import React from 'react';
import { shallow } from 'enzyme';
import { BoardComponent as Board } from './Board.component';

describe('Board', () => {

  beforeEach(() => {
  })

  test('should render Board', () => {
    const squares = []
    expect(shallow(<Board process={ { squares } } />)).toMatchSnapshot();
  })
})
