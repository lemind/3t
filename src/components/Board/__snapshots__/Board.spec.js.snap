// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Board should render Board 1`] = `
<div
  className="board"
>
  <div
    className="board__row"
  >
    <SquareComponent
      onClick={[Function]}
    />
    <SquareComponent
      onClick={[Function]}
    />
    <SquareComponent
      onClick={[Function]}
    />
  </div>
  <div
    className="board__row"
  >
    <SquareComponent
      onClick={[Function]}
    />
    <SquareComponent
      onClick={[Function]}
    />
    <SquareComponent
      onClick={[Function]}
    />
  </div>
  <div
    className="board__row"
  >
    <SquareComponent
      onClick={[Function]}
    />
    <SquareComponent
      onClick={[Function]}
    />
    <SquareComponent
      onClick={[Function]}
    />
  </div>
</div>
`;
